### DISCUSSION TITLE
--reinstall | numpy error in log, but automtic working despite the error

### DISCUSSION THREAD

**Discussion by cgidesign-de** [vladmandic/sdnext#960] [Q&A]
Hi,
I used --reinstall flag
completed with one error
permission issue with one dll
I checked the permissions - user has full control (read, write, change)

Log (shortened some entries):
... | ERROR | F:\automatic\installer.py | Error running pip: install --upgrade numpy
... | DEBUG | F:\automatic\installer.py | Pip output: Requirement already satisfied: numpy in f:\automatic\venv\lib\site-packages (1.23.5)
Collecting numpy
  Downloading numpy-1.24.3-cp310-cp310-win_amd64.whl (14.8 MB)
     --------------------------------------- 14.8/14.8 MB 29.8 MB/s eta 0:00:00
Installing collected packages: numpy
  Attempting uninstall: numpy
    Found existing installation: numpy 1.23.5
    Uninstalling numpy-1.23.5:
      Successfully uninstalled numpy-1.23.5

**ERROR: Could not install packages due to an OSError: [WinError 5] Access is denied: 'F:\\...\\.libs\\libopenblas.FB5AE2TYXYH2IJRDKGDGQ3XBKLKTF43H.gfortran-win_amd64.dll'

Check the permissions.**

But automatics works despite the error.

Should I just ignore the error?

**Comment by vladmandic** [comment]
ah, magic of python on windows. root cause is messed up and i hate python for not addressing it a looooong time ago.
basically, on linux filesystem only atomic operation (operation that is guaranteed not to run into any possible sharing violation like some other process modified the file while you're using it, etc.) is rename. file locking doesn't exist. so during installation python writes stuff as temp files and once its done with everything, it renames them to actual desired name.

but windows works differently - it actually has a file lock. so running create/write/rename quickly can run into access violation.

no harm since old version never got actually removed. and typically retrying works as its a timing thing, not an actual error.

**Comment by cgidesign-de** [comment]
Thank you for the explanation. It was solved after a subsequent reinstall.
